Add helm chart

mehyedes opened the pull request:
This adds a helm chart for `cluster-api-state-metrics`

- [x] Create Helm chart
- [x] Add make targets for helm
- [ ] ~~Set default image repo (related to https://github.com/mercedes-benz/cluster-api-state-metrics/issues/3)~~
- [ ] ~~Add additional CI steps for the helm chart~~ Will be done in a separate PR
- [ ] ~~Set helm repo for publishing the chart~~ Will be done in a separate PR

It would be great if you provide some feedback on how to proceed with the points above :grin: 

In the chart, there are also [serviceMonitor](https://prometheus-operator.dev/docs/operator/api/#servicemonitor) and [GrafanaDashboard](https://github.com/grafana-operator/grafana-operator/blob/master/documentation/api.md#grafanadashboard) templates. We thought they might be useful for anyone out there using the prometheus and grafana operators additionally.

Solves: https://github.com/mercedes-benz/cluster-api-state-metrics/issues/9

Commit:
add helm chart

Added first version 0.1.0 of the cluster-api-state-metrics helm chart.

Signed-off-by: Mehdi Yedes <mehdi.yedes@telekom.de>

Commit:
add make targets for helm

Signed-off-by: Mehdi Yedes <mehdi.yedes@telekom.de>

dimatha commented on deploy/chart/templates/deployment.yaml:
```diff
@@ -0,0 +1,155 @@
+            httpGet:
+              path: /healthz
+              port: telemetry
+              initialDelaySeconds: 15
```
Indentation

dimatha commented on deploy/chart/templates/deployment.yaml:
```diff
@@ -0,0 +1,155 @@
+            httpGet:
+              path: /readyz
+              port: metrics
+              initialDelaySeconds: 5
```
Indentation

chrischdi commented:
Looks good so far :grinning: thanks for your Pull Request :tada: 

Regarding your points:

> Add additional CI steps for the helm chart

The rbac role is getting generated and updated via kubebuilder.

Should we integrate something to the makefile to also update the rbac role in the helm chart? (sed or something like that to generate the helm variant from the kubebuilder generated one?)

> Set default image repo (related to https://github.com/mercedes-benz/cluster-api-state-metrics/issues/3)

I hope that I'll be able to finish the process by next week so we can publish images to quay.io.

> Set helm repo for publishing the chart

What do we have for options here?
Is a github release a valid source or could this also be done at quay.io then?

chrischdi commented on deploy/chart/templates/grafanaDashboard.yaml:
```diff
@@ -0,0 +1,2346 @@
+            "value": "a4-1"
+            },
+            "datasource": null,
+            "definition": "label_values(capi_cluster_labels{exported_namespace=~\"$namespace\"}, cluster)",
```
these queries do not match for everyone. Should be `capi_kubeadmcontrolplane_labels`?!

mehyedes replied:
> these queries do not match for everyone

I am not sure I fully understand what you mean by that :thinking: 

This query is only for getting the list of the clusters and use it for the `cluster` variable. We could also get the same result by using the query below as the kubeadmControlPlane has the same name as the cluster that owns it:
```
label_values(kubeadmcontrolplane_cluster_labels{exported_namespace=~\"$namespace\"}, kubeadmcontrolplane)
```

apricote replied:
Some clusters might not use `KubeadmControlPlane` but instead some other ControlPlane provider such as `Talos`.

mehyedes replied:
True, `KubeadmControlPlane` is not an option here. At least for now.

chrischdi replied:
Ah yes, sorry looks like I misread something here. 

However, let's discuss the dashboards in a separate PRs.

chrischdi reviewed (commented):
I think we should split out the grafana dashboard to either a separate location or another PR.

Not everyone is using the `integreatly.org/v1alpha1` CRs and the dashboard would be way more reusable then I think.

Commit:
update helm chart templates

fixed indentation in deployment template
removed grafanaDashboard template
updated chart README

Signed-off-by: Mehdi Yedes <mehdi.yedes@telekom.de>

mehyedes commented:
> I think we should split out the grafana dashboard to either a separate location or another PR.
> 
> Not everyone is using the `integreatly.org/v1alpha1` CRs and the dashboard would be way more reusable then I think.

Agree, would be better to leave it out of this PR for now and then do it in another iteration.

mehyedes commented:
> > Add additional CI steps for the helm chart
> 
> The rbac role is getting generated and updated via kubebuilder.
> 
> Should we integrate something to the makefile to also update the rbac role in the helm chart? (sed or something like that to generate the helm variant from the kubebuilder generated one?)

Yes, that would make sense. However, we should keep in mind that we need to publish a new helm chart release whenever the role(s) are changed. Of course, that would be ideally automated but we just need to keep that in mind.
 
> > Set default image repo (related to #3)
> 
> I hope that I'll be able to finish the process by next week so we can publish images to quay.io.

Awesome! 

> > Set helm repo for publishing the chart
> 
> What do we have for options here? Is a github release a valid source or could this also be done at quay.io then?

I think Quay already supports helm chart repos(never worked with it though). But simply using github releases might also be an option for now. We could use [chart-releaser](https://github.com/helm/chart-releaser), but I am not sure if it's possible to use it for this repo, since it's not dedicated to only storing helm charts.

chrischdi commented:
> > > Add additional CI steps for the helm chart
> > 
> > 
> > The rbac role is getting generated and updated via kubebuilder.
> > Should we integrate something to the makefile to also update the rbac role in the helm chart? (sed or something like that to generate the helm variant from the kubebuilder generated one?)
> 
> Yes, that would make sense. However, we should keep in mind that we need to publish a new helm chart release whenever the role(s) are changed. Of course, that would be ideally automated but we just need to keep that in mind.
> 

I think publishing a chart release should be fine using chart releaser's github action: https://github.com/helm/chart-releaser-action :-) but let's do this in a follow-up PR then to limit scope here.

> > > Set default image repo (related to #3)
> > 
> > 
> > I hope that I'll be able to finish the process by next week so we can publish images to quay.io.
> 
> Awesome!
> 
> > > Set helm repo for publishing the chart
> > 
> > 
> > What do we have for options here? Is a github release a valid source or could this also be done at quay.io then?
> 
> I think Quay already supports helm chart repos(never worked with it though). But simply using github releases might also be an option for now. We could use [chart-releaser](https://github.com/helm/chart-releaser), but I am not sure if it's possible to use it for this repo, since it's not dedicated to only storing helm charts.

What would you prefer as chart repo: github release or quay.io? I think github release would be better?!

mehyedes commented:
> What would you prefer as chart repo: github release or quay.io? I think github release would be better?!

I agree. 

I'll try to update my PR in the upcoming few days.

Commit:
update helm chart

- Added script to generate role templates for the helm chart. The script is executed with `make
  manifests`
- Added missing SPDX headers

Signed-off-by: Mehdi Yedes <mehdi.yedes@telekom.de>

mehyedes commented:
I have updated the PR now. 

The failing checks should be resolved and the role templates for the helm chart are now generated from the ones under `config/rbac`

chrischdi commented:
Hi @mehyedes sorry for the late reply.

We are currently working on a proposal to contribute CASM to github.com/kubernetes-sigs/cluster-api .
I don't know if the contribution would include a helm chart if we have it in here.

So I see some options here:

* Add the helm chart here and also further maintain it after the CASM code got contributed to cluster-api
* Create a seperate repository for the helm chart (which could be also published directly by you if you would prefer that), this way the chart would be independent.
* Later on maybe contribute a helm chart to github.com/kubernetes-sigs/cluster-api if they want/accept helm-charts there (e.g. somewhere in `hack/observability`)

What do you think?

mehyedes commented:
Hey @chrischdi, thanks for your feedback.

I think it depends on how the contribution of CASM into cluster-api would be. If CASM will be integrated into the cluster-api codebase, I believe there would be no need for a separate helm chart at all in that case. AFAIK, there is no "official" helm chart for cluster-api for now, but if there was one, then that's where I would add the CASM config.
And since CAPI project is using `clusterctl` to install its components, I don't see that there will be a helm chart, at least not any time soon.
Maybe we could add the helm chart into this repo, until the project gets contributed to CAPI. Then, there will be no need for it anymore I think so it wouldn't need to be further maintained.

chrischdi commented on Makefile:
```diff
@@ -67,6 +68,9 @@ vet: ## Run go vet against code.
 lint: # Run golangci-lint against code.
 	@hack/check_golangci-lint.sh
 
+helm-lint: # Run helm lint against helm chart code
```
Maybe makes sense to add `helm-lint` to the `test` target in line 91?

tobiasgiese replied:
Or to the `lint` target in l.68

chrischdi replied:
as tobias wrote: would be better at the `lint` target

mehyedes replied:
done

chrischdi reviewed (commented):
one nit

/lgtm

@tobiasgiese maybe you have some time to take a look at it too?

chrischdi commented:
> Maybe we could add the helm chart into this repo, until the project gets contributed to CAPI. Then, there will be no need for it anymore I think so it wouldn't need to be further maintained.

Sounds fair to me :+1: Just didn't want to raise the expectation that it will be integrated to the core repo automatically :-) So we have the same opinion here :smile:

tobiasgiese commented on deploy/chart/templates/tests/test-connection.yaml:
```diff
@@ -0,0 +1,15 @@
+    - name: wget
+      image: busybox
+      command: ['wget']
+      args: ['{{ include "cluster-api-state-metrics.fullname" . }}:{{ .Values.service.port }}']
```
```suggestion
      command:
        - "wget"
      args:
        - "{{ include "cluster-api-state-metrics.fullname" . }}:{{ .Values.service.port }}"
```

I think it's better to use yaml instead of json here

tobiasgiese reviewed (approved):
One tiny nit, rest looks really fine to me. Thanks for your contribution! 🙂🎉

seanschneeweiss reviewed (commented):
/lgtm

seanschneeweiss reviewed (approved):
Thank you a lot for the contribution.

Commit:
update Makefile

Added helm-lint to the lint target

Signed-off-by: Mehdi Yedes <mehdi.yedes@telekom.de>

mehyedes committed:
Update deploy/chart/templates/tests/test-connection.yaml

Use YAML instead JSON

Co-authored-by: Tobias Giese <tobiasgiese@gmail.com>

johannesfrey reviewed (approved):
Awesome. Thank you very much.

chrischdi commented:
One last nit, the Readme seems not to be happy :thinking:

Commit:
update README

Updated make targets docs

chrischdi commented:
Thank you very much for your contribution 👍 If I remember correctly: the first external contribution for a Mercedes Benz repo on GitHub 🎉
